ImportError: dlopen(…/kerberos.cpython-36m-darwin.so, 2): Symbol not found

I'm on macOS X 10.13.2 (High Sierra), Python 3.6.5 installed via MacPorts, and `pykerberos` 1.2.1 installed via pip:

```
$ pip-3.6 show pykerberos
Name: pykerberos
Version: 1.2.1
Summary: High-level interface to Kerberos
Home-page: UNKNOWN
Author: UNKNOWN
Author-email: UNKNOWN
License: ASL 2.0
Location: /opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages
Requires:
Required-by:
```

I'm unable to `import kerberos`, however:

```
$ python3.6
Python 3.6.5 (default, Mar 29 2018, 15:37:32)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import kerberos
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen(/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/kerberos.cpython-36m-darwin.so, 2): Symbol not found: _mempcpy
  Referenced from: /opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/kerberos.cpython-36m-darwin.so
  Expected in: flat namespace
 in /opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/kerberos.cpython-36m-darwin.so
```

I don't even know how to debug this. Help!

A manual build via `python3.6 setup.py build` doesn't work either. Here's the build log:

```
joule:~/tmp/pykerberos @master u=> python3.6 setup.py build
running build
running build_ext
building 'kerberos' extension
creating build
creating build/temp.macosx-10.13-x86_64-3.6
creating build/temp.macosx-10.13-x86_64-3.6/src
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberos.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberos.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberosbasic.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberosbasic.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberosgss.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberosgss.o -I/opt/local/include -DGSSAPI_EXT
src/kerberosgss.c:402:29: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
        if (len-4-token_len < 0)
            ~~~~~~~~~~~~~~~ ^ ~
src/kerberosgss.c:876:14: warning: implicit declaration of function 'mempcpy' is invalid in C99 [-Wimplicit-function-declaration]
    outloc = mempcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
             ^
src/kerberosgss.c:876:14: warning: this function declaration is not a prototype [-Wstrict-prototypes]
src/kerberosgss.c:876:12: warning: incompatible integer to pointer conversion assigning to 'char *' from 'int' [-Wint-conversion]
    outloc = mempcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
           ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 warnings generated.
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberospw.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberospw.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/base64.c -o build/temp.macosx-10.13-x86_64-3.6/src/base64.o -I/opt/local/include -DGSSAPI_EXT
creating build/lib.macosx-10.13-x86_64-3.6
/usr/bin/clang -bundle -undefined dynamic_lookup -L/opt/local/lib -Wl,-headerpad_max_install_names build/temp.macosx-10.13-x86_64-3.6/src/kerberos.o build/temp.macosx-10.13-x86_64-3.6/src/kerberosbasic.o build/temp.macosx-10.13-x86_64-3.6/src/kerberosgss.o build/temp.macosx-10.13-x86_64-3.6/src/kerberospw.o build/temp.macosx-10.13-x86_64-3.6/src/base64.o -o build/lib.macosx-10.13-x86_64-3.6/kerberos.cpython-36m-darwin.so -L/opt/local/lib -dynamic -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err
```

Sorry for sleeping on this so long. Work was too hectic, and I actually don't have much to clarify. When I build manually, install, and `import`, I get the same error message as originally reported:

```
joule:~/tmp/pykerberos @master u=> virtualenv-3.6 ~/pyvenv/pykerberos-py3.6
Using base prefix '/opt/local/Library/Frameworks/Python.framework/Versions/3.6'
New python executable in /Users/julian/pyvenv/pykerberos-py3.6/bin/python3.6
Also creating executable in /Users/julian/pyvenv/pykerberos-py3.6/bin/python
Installing setuptools, pip, wheel...done.

joule:~/tmp/pykerberos @master u=> . ~/pyvenv/pykerberos-py3.6/bin/activate

(pykerberos-py3.6) joule:~/tmp/pykerberos @master u=> python setup.py build
running build
running build_ext
building 'kerberos' extension
creating build
creating build/temp.macosx-10.13-x86_64-3.6
creating build/temp.macosx-10.13-x86_64-3.6/src
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberos.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberos.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberosbasic.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberosbasic.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberosgss.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberosgss.o -I/opt/local/include -DGSSAPI_EXT
src/kerberosgss.c:402:29: warning: comparison of unsigned expression < 0 is always false
      [-Wtautological-compare]
        if (len-4-token_len < 0)
            ~~~~~~~~~~~~~~~ ^ ~
src/kerberosgss.c:876:14: warning: implicit declaration of function 'mempcpy' is invalid in C99
      [-Wimplicit-function-declaration]
    outloc = mempcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
             ^
src/kerberosgss.c:876:14: warning: this function declaration is not a prototype
      [-Wstrict-prototypes]
src/kerberosgss.c:876:12: warning: incompatible integer to pointer conversion assigning to 'char *'
      from 'int' [-Wint-conversion]
    outloc = mempcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
           ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 warnings generated.
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/kerberospw.c -o build/temp.macosx-10.13-x86_64-3.6/src/kerberospw.o -I/opt/local/include -DGSSAPI_EXT
/usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -pipe -Os -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c src/base64.c -o build/temp.macosx-10.13-x86_64-3.6/src/base64.o -I/opt/local/include -DGSSAPI_EXT
creating build/lib.macosx-10.13-x86_64-3.6
/usr/bin/clang -bundle -undefined dynamic_lookup -L/opt/local/lib -Wl,-headerpad_max_install_names build/temp.macosx-10.13-x86_64-3.6/src/kerberos.o build/temp.macosx-10.13-x86_64-3.6/src/kerberosbasic.o build/temp.macosx-10.13-x86_64-3.6/src/kerberosgss.o build/temp.macosx-10.13-x86_64-3.6/src/kerberospw.o build/temp.macosx-10.13-x86_64-3.6/src/base64.o -o build/lib.macosx-10.13-x86_64-3.6/kerberos.cpython-36m-darwin.so -L/opt/local/lib -dynamic -lgssapi_krb5 -lkrb5 -lk5crypto -lcom_err

(pykerberos-py3.6) joule:~/tmp/pykerberos @master u=> python setup.py install
running install
running bdist_egg
running egg_info
creating pykerberos.egg-info
writing pykerberos.egg-info/PKG-INFO
writing dependency_links to pykerberos.egg-info/dependency_links.txt
writing top-level names to pykerberos.egg-info/top_level.txt
writing manifest file 'pykerberos.egg-info/SOURCES.txt'
reading manifest file 'pykerberos.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'pykerberos.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-10.13-x86_64/egg
running install_lib
running build_ext
creating build/bdist.macosx-10.13-x86_64
creating build/bdist.macosx-10.13-x86_64/egg
copying build/lib.macosx-10.13-x86_64-3.6/kerberos.cpython-36m-darwin.so -> build/bdist.macosx-10.13-x86_64/egg
creating stub loader for kerberos.cpython-36m-darwin.so
byte-compiling build/bdist.macosx-10.13-x86_64/egg/kerberos.py to kerberos.cpython-36.pyc
creating build/bdist.macosx-10.13-x86_64/egg/EGG-INFO
copying pykerberos.egg-info/PKG-INFO -> build/bdist.macosx-10.13-x86_64/egg/EGG-INFO
copying pykerberos.egg-info/SOURCES.txt -> build/bdist.macosx-10.13-x86_64/egg/EGG-INFO
copying pykerberos.egg-info/dependency_links.txt -> build/bdist.macosx-10.13-x86_64/egg/EGG-INFO
copying pykerberos.egg-info/top_level.txt -> build/bdist.macosx-10.13-x86_64/egg/EGG-INFO
writing build/bdist.macosx-10.13-x86_64/egg/EGG-INFO/native_libs.txt
zip_safe flag not set; analyzing archive contents...
__pycache__.kerberos.cpython-36: module references __file__
creating dist
creating 'dist/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg' and adding 'build/bdist.macosx-10.13-x86_64/egg' to it
removing 'build/bdist.macosx-10.13-x86_64/egg' (and everything under it)
Processing pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg
creating /Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg
Extracting pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg to /Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages
Adding pykerberos 1.2.2.dev0 to easy-install.pth file

Installed /Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg
Processing dependencies for pykerberos==1.2.2.dev0
Finished processing dependencies for pykerberos==1.2.2.dev0

(pykerberos-py3.6) joule:~/tmp/pykerberos @master u=> python
Python 3.6.5 (default, Mar 29 2018, 15:37:32)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import kerberos
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen(/Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg/kerberos.cpython-36m-darwin.so, 2): Symbol not found: _mempcpy
  Referenced from: /Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg/kerberos.cpython-36m-darwin.so
  Expected in: flat namespace
 in /Users/julian/pyvenv/pykerberos-py3.6/lib/python3.6/site-packages/pykerberos-1.2.2.dev0-py3.6-macosx-10.13-x86_64.egg/kerberos.cpython-36m-darwin.so
```

--- a/src/kerberosgss.c
+++ b/src/kerberosgss.c
@@ -873,9 +873,10 @@ int encrypt_message(gss_client_state *state, char *message_input, char **header,
     *encrypted_data_len = iov[1].buffer.length + iov[2].buffer.length;
     *encrypted_data = malloc(*encrypted_data_len);
     outloc = *encrypted_data;
-    outloc = mempcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
+    memcpy(outloc, iov[1].buffer.value, iov[1].buffer.length);
+    outloc += iov[1].buffer.length;
     // NB: no-op if no padding is necessary (which seems to always be the case with aes256-cts-hmac-sha1-96)
-    mempcpy(outloc, iov[2].buffer.value, iov[2].buffer.length);
+    memcpy(outloc, iov[2].buffer.value, iov[2].buffer.length);
 
     ret = 0;
 end:

